pkg/controller/migmigration/verify.go

```diff
@@ -0,0 +1,329 @@
+	VerificationProgressing
+	// VerificationSucceeded is a successfull validation state
+	VerificationSucceeded
+)
+
+const appVerificationFailed = "AppVerificationFailed"
+
+// CrashLoopBackOffMargin is a CrashLoopBackOff margin
+const CrashLoopBackOffMargin = 3
+
+var podManagingResources = []schema.GroupVersionKind{
+	schema.GroupVersionKind{
+		Group: "apps",
+		Kind:  "statefulset",
+	},
+	schema.GroupVersionKind{
+		Group: "apps",
+		Kind:  "deployment",
+	},
+	schema.GroupVersionKind{
+		Group: "extensions",
+		Kind:  "replicaset",
+	},
+	schema.GroupVersionKind{
+		Group: "apps.openshift.io",
+		Kind:  "deploymentconfig",
+	},
+}
+
+func (t *Task) verifyAppState() (int, error) {
+	// Scan namespaces
+	plan := t.PlanResources.MigPlan
+	for _, namespace := range plan.Spec.Namespaces {
+		result, err := t.verifyInNamespaceHealth(namespace)
+		if err != nil {
+			return VerificationFailed, err
+		}
+		if result != VerificationSucceeded {
+			return result, nil
+		}
+	}
+	return VerificationSucceeded, nil
+}
+
+func (t *Task) verifyInNamespaceHealth(namespace string) (int, error) {
+	options := k8sclient.InNamespace(namespace)
+	client, err := t.getDestinationClient()
+	if err != nil {
+		return VerificationFailed, err
+	}
+
+	// verify Jobs
+	healthy, err := t.verifyJobs(client, options)
+	if err != nil {
+		return VerificationFailed, err
+	}
+	if !healthy {
+		return VerificationFailed, nil
+	}
+
+	// verify DaemonSets
+	validationResult, err := t.verifyDaemonSets(client, options)
+	if err != nil {
+		return VerificationFailed, err
+	}
+	if validationResult != VerificationSucceeded {
+		return validationResult, nil
+	}
+
+	// verify generic resources
+	return t.verifyOwnerResources(client, options)
+}
+
+// verify pod owner resources includes ReplicaSets, Deployments, etc.
+func (t *Task) verifyOwnerResources(client k8sclient.Client, options *k8sclient.ListOptions) (int, error) {
+	podManagersList := unstructured.UnstructuredList{}
+
+	for _, gvk := range podManagingResources {
+		podManagersList.SetGroupVersionKind(gvk)
+		err := client.List(context.TODO(), options, &podManagersList)
+		if err != nil {
+			return VerificationFailed, err
+		}
+
+		for _, podManager := range podManagersList.Items {
+			expectedReplicas, found, err := unstructured.NestedFieldCopy(podManager.Object, "spec", "replicas")
+			if err != nil {
+				return VerificationFailed, err
+			}
+			if !found {
+				err = fmt.Errorf("Replicas field was not found in %s kind", podManager.GetKind())
+				return VerificationFailed, err
+			}
+			// Skip downscaled owner resources
+			if expectedReplicas == 0 {
+				continue
+			}
+
+			readyReplicas, found, err := unstructured.NestedFieldCopy(podManager.Object, "status", "readyReplicas")
+			if err != nil {
+				return VerificationFailed, err
+			}
+			// Skip resource with uprepared status
+			if !found {
+				_, found, err := unstructured.NestedFieldCopy(podManager.Object, "status", "unavailableReplicas")
```

Comment:
This block feels weird... So if the `readyReplicas` field isn't found, we then check for `unavailableReplicas`. If `unavailableReplicas` aren't found, we return an error that `replicas field was not found in kind`.

Do we need two conditionals here? Won't `err` be set if `NestedFieldCopy` fails when `unavailableReplicas` doesn't exist? If that's the case we can remove the `if !found` block on L140. If the error message is nonsensical then I can understand having the two conditions here.

If `unavailableReplicas` are found I'm assuming we are continuing to get the failing pods health?

Reply:
After some investigation, I removed this part. Just looking into the `Pods` is enough. The error suggests only that there was some non string accessor in the structure, so it does not give anything. And another thing - the `unavailableReplicas` is also for pods in `Pending` state :(